compiler/optimizer/abstractinterpreter/IDTNode.cpp

```diff
@@ -0,0 +1,172 @@
+      int32_t idx, 
+      TR_CallTarget* callTarget,
+      TR::ResolvedMethodSymbol* symbol,
+      uint32_t byteCodeIndex, 
+      float callRatio, 
+      IDTNode *parent, 
+      uint32_t budget) :
+   _idx(idx),
+   _callTarget(callTarget),
+   _staticBenefit(0),
+   _byteCodeIndex(byteCodeIndex),
+   _symbol(symbol),
+   _children(NULL),
+   _parent(parent),
+   _budget(budget),
+   _callRatio(callRatio),
+   _rootCallRatio(parent ? parent->_rootCallRatio * callRatio : 1),
+   _inliningMethodSummary(NULL)
+   {   
+   }
+
+TR::IDTNode* TR::IDTNode::addChild(
+      int32_t idx,
+      TR_CallTarget* callTarget,
+      TR::ResolvedMethodSymbol* symbol,
+      uint32_t byteCodeIndex, 
+      float callRatio,
+      TR::Region& region)
+   {
+   uint32_t budget =  getBudget() - callTarget->_calleeMethod->maxBytecodeIndex();
+   
+   TR::IDTNode* newChild = new (region) TR::IDTNode(
+                                             idx, 
+                                             callTarget,
+                                             symbol,
+                                             byteCodeIndex, 
+                                             callRatio,
+                                             this,
+                                             budget);
+
+   // The case where there is no children
+   if (getNumChildren() == 0)
+      {
+      setOnlyChild(newChild);
+      return newChild;
+      }   
+
+   // The case where there is 1 child
+   if (getNumChildren() == 1)
+      {
+      TR::IDTNode* onlyChild = getOnlyChild();
+      _children = new (region) TR::vector<TR::IDTNode*, TR::Region&>(region); 
+      _children->push_back(onlyChild);
+      }
+                     
+   _children->push_back(newChild);
+   return _children->back();
+   }
+
+uint32_t TR::IDTNode::getNumDescendants()
+   {
+   const uint32_t numChildren = getNumChildren();
+   uint32_t sum = 0;
+   for (uint32_t i =0; i < numChildren; i ++)
+      {
+      sum += 1 + getChild(i)->getNumDescendants(); 
+      }
+   return sum;
+   }
+
+uint32_t TR::IDTNode::getRecursiveCost() 
+   {
+   const uint32_t numChildren = getNumChildren();
+   uint32_t cost = getCost();
+   for (uint32_t i = 0; i < numChildren; i ++)
+      {
+      IDTNode *child = getChild(i);
+      cost += child->getRecursiveCost();
+      }
+   
+   return cost;
+   }
+
+uint32_t TR::IDTNode::getNumChildren()
+   {
+   if (_children == NULL)
+      return 0;
+   
+   if (getOnlyChild() != NULL)
+      return 1;
+   
+   size_t num = _children->size();
+   return num;
+   }
+
+TR::IDTNode* TR::IDTNode::getChild(uint32_t index)
+   {
+   const uint32_t numChildren = getNumChildren();
+   TR_ASSERT_FATAL(index < numChildren, "Child index out of range!\n");
+
+   if (index == 0 && numChildren == 1) // only one child
+      return getOnlyChild();
+
+   return (*_children)[index];
+   }
+
+uint64_t TR::IDTNode::getBenefit()
+   {
+   return _rootCallRatio  * (1 + _staticBenefit) * 10;
```

Comment:
It might be good to have a comment for what's going on in here. I'm guessing:
- `1 +` just signifies there is a performance benefit from eliminating the call
- `* 10` gives us some resolution when `_rootCallRatio` starts getting small

Also, this appears to be implicitly casting the product from `double` to `uint64_t`. Given my guess above about `* 10`, I think that's intentional, but maybe it should be an explicit cast

Reply:
The 1 is used to prevent multiplying with 0.  The 10 is the equal weight for every benefit.

Reply:
Ok, would you mind writing a comment about it in the code? (Rather than a GitHub comment - I suppose I should have been more specific :sweat_smile:)

And what do you think about the cast?

Reply:
Thanks for the code comment! What about the cast?

Reply:
Reflecting our conversation here: The return type has been changed to `double`, and I just hadn't noticed the change